### PR TITLE
Sm/event-task-custom-fields

### PR DESCRIPTION
**Pull request by mshanemc**
Event/Task in profiles now match Activity in (package.xml).

rewrite of the Profile Rewrite process
- it's now extracted as a testable function.  Pass in a profile (in JSON) and a map representation of the package.xml contents, get back what the profile would be
- uses type-safe Profile metadata information from jsforce
- those types allow manipulation of json properties instead of using DOM tooling to deal with XML
- moved logging of Profile modifications from a separate ProfileInformation class to inline (let logger handle level decisions
- removed some unused dependencies

https://github.com/forcedotcom/cli/issues/2278
@W-13753544@

### CODE REVIEW

**Review by shetzel — CHANGES_REQUESTED** [review]
Nice!  Requested some small changes.

File: src/package/packageProfileApi.ts [hunk 1]
```diff
@@ -108,112 +49,46 @@ export class PackageProfileApi extends AsyncCreatable<ProfileApiOptions> {
...
-            // It is normal for the file to not exist if the profile is in the workspace but not in the directory being packaged.
-            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-            if (err instanceof Error && 'code' in err && err.code !== 'ENOENT') {
-              throw err;
-            }
-          }
-        }
-      }
-    });
+    const excludedProfiles = results
```
Comment (WillieRuemmele) [inline src/package/packageProfileApi.ts:83]: NIT: inline and `return results.filter()...`

File: src/package/packageVersionCreate.ts [hunk 2]
```diff
@@ -464,19 +464,17 @@ export class PackageVersionCreate {
     // metadata exclusions. If necessary, read the existing package.xml and then re-write it.
     const currentPackageXml = await fs.promises.readFile(path.join(packageVersMetadataFolder, 'package.xml'), 'utf8');
     // convert to json
-    const packageJson = (await xml2js.parseStringPromise(currentPackageXml)) as {
-      Package: { types: Array<{ name: string[]; members: string[] }>; version: string };
-    };
-    if (!packageJson?.Package) {
+    const packageXmlAsJson = (await xml2js.parseStringPromise(currentPackageXml)) as PackageXml;
```
Comment (WillieRuemmele) [inline src/package/packageVersionCreate.ts:467]: not directly related, but we can remove the `clientSideInfo` param L459

File: src/package/packageProfileApi.ts [hunk 3]
```diff
@@ -225,117 +100,69 @@ export class PackageProfileApi extends AsyncCreatable<ProfileApiOptions> {
...
+
+const getXmlFileLocation = (destPath: string, profilePath: string): string =>
+  path.join(destPath, path.basename(profilePath).replace(/(.*)(-meta.xml)/, '$1'));
+
+const deleteButAllowEnoent = (destFilePath: string): void => {
+  try {
+    fs.unlinkSync(destFilePath);
+  } catch (err) {
+    // It is normal for the file to not exist if the profile is in the workspace but not in the directory being packaged.
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
```
Comment (WillieRuemmele) [inline src/package/packageProfileApi.ts:157]: what's being unsafely accessed? might not need this anymore

**Review by WillieRuemmele — COMMENTED** [review]
Agree with Steve's comments

File: src/interfaces/packagingInterfacesAndType.ts [hunk 4]
```diff
@@ -469,3 +468,10 @@ export const Package1VersionEvents = {
     progress: 'Package1Version/create-progress',
   },
 };
+
+export type PackageXml = {
+  Package: {
+    types: Array<{ name: string; members: string[] }>;
+    version: string;
+  };
+};
```
[the remark this answers is not in the document]
Comment (mshanemc) [reply src/interfaces/packagingInterfacesAndType.ts:477]: defined, yes.  Exported, not at the top-level.

I'm going to get it from jsforce

File: src/package/packageProfileApi.ts [hunk 5]
```diff
@@ -108,112 +49,46 @@ export class PackageProfileApi extends AsyncCreatable<ProfileApiOptions> {
    */
   public generateProfiles(
     destPath: string,
-    manifest: {
-      Package: Array<{ name: string[]; members: string[] }>;
-    },
+    manifestTypes: PackageXml['Package']['types'],
```
[the remark this answers is not in the document]
Comment (mshanemc) [reply src/package/packageProfileApi.ts:52]: done

File: src/package/packageVersionCreate.ts (hunk 2 above)
Comment (mshanemc) [reply src/package/packageVersionCreate.ts:467]: done. all destructured because 11 params is hard to manage

**Review by WillieRuemmele — APPROVED** [review]

File: src/package/packageVersionCreate.ts [hunk 6]
```diff
@@ -1143,3 +1127,41 @@ export class MetadataResolver {
...
+export const packageXmlStringToPackageXmlJson = (rawXml: string): PackageXml => {
+  const parser = new XMLParser({
+    ignoreAttributes: true,
+    parseTagValue: false,
+    parseAttributeValue: false,
+    cdataPropName: '__cdata',
+    ignoreDeclaration: true,
+    numberParseOptions: { leadingZeros: false, hex: false },
+    // make sure types and members is always an array
+    isArray: (name: string) => ['types', 'members'].includes(name),
```
Comment (WillieRuemmele) [inline src/package/packageVersionCreate.ts:1140]: nice they provide this

**Review by WillieRuemmele — APPROVED** [review]

File: src/package/profileRewriter.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    props.filter((item) => packageXml.get('CustomField')?.includes(fieldCorrections(item.field))),
+
+  layoutAssignments: (props: RewriteProps['layoutAssignments'], packageXml: PackageMap) =>
+    props.filter((item) => packageXml.get('Layout')?.includes(item.layout)),
+
+  tabVisibilities: (props: RewriteProps['tabVisibilities'], packageXml: PackageMap) =>
+    props.filter((item) => packageXml.get('CustomTab')?.includes(item.tab)),
+
+  applicationVisibilities: (props: RewriteProps['applicationVisibilities'], packageXml: PackageMap) =>
+    props.filter((item) => packageXml.get('Application')?.includes(item.application)),
```
Comment (amtrack) [inline src/package/profileRewriter.ts:96]:
```suggestion
    props.filter((item) => packageXml.get('CustomApplication')?.includes(item.application)),
```

**Review by amtrack — CHANGES_REQUESTED** [review]
`applicationVisibilities` were removed incorrectly:

```
The "userPermissions" profile setting was removed from the "REDACTED Profile" profile.
```

The metadata type is `CustomApplication` not `Application`.

**Pull request merged**